File: wp.m
@import Foundation;

#define BUFFER_SIZE 8192

@implementation NSMutableDictionary(NSMutableDictionaryExt)

- (void)incrementNumberForKey:(NSString *)key {
    NSNumber *count = [self objectForKey:key];
    count = (count) ? [NSNumber numberWithInt:[count intValue] + 1]
                    : [NSNumber numberWithInt:1];
    [self setObject:count forKey:key];
}

@end

@implementation NSFileHandle(NSFileHandleExt)

- (NSString *)readLine {
    char buffer[BUFFER_SIZE];
    return fgets(buffer, sizeof buffer, stdin)
        ? [NSString stringWithCString:buffer encoding:NSASCIIStringEncoding]
            : @"";
}

@end

int main(void) {
    @autoreleasepool {

        NSMutableDictionary *wordsDict = [[NSMutableDictionary alloc] init];
        NSString *line;
        NSFileHandle *stdin = [NSFileHandle fileHandleWithStandardInput];

        do {
            line = [stdin readLine];
            NSArray *words = [[line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] componentsSeparatedByString:@" "];

            [words enumerateObjectsUsingBlock:^(id word, __unused NSUInteger idx, __unused BOOL *stop) {
                if ([word length] > 0)
                    [wordsDict incrementNumberForKey:word];
            }];

        } while ([line length] > 0);

        NSArray *wordList = [wordsDict keysSortedByValueUsingSelector:@selector(compare:)];
        [wordList enumerateObjectsWithOptions:NSEnumerationReverse usingBlock:^(id word, __unused NSUInteger idx, __unused BOOL *stop) {
            printf("%lu %s\n", (unsigned long)[[wordsDict objectForKey:word] intValue], [word cString]);
        }];

    }

    return 0;
}

File: wp.swift
import Foundation

let bufferSize = 8192

func readline(fd: CMutablePointer<FILE>) -> String {
  var buffer = CChar[](count: bufferSize, repeatedValue: 0)

  fgets(&buffer, CInt(bufferSize), fd)

  var len = 0
  for i in 0..bufferSize {
    if buffer[i] == 0x0a {
      len = i + 1
      break
    }
  }

  return NSString(bytes:buffer, length:len, encoding:NSUTF8StringEncoding)
}

let stdin = NSFileHandle.fileHandleWithStandardInput()
let fd: CMutablePointer<FILE> = fdopen(stdin.fileDescriptor, "r")

var wordsDict = Dictionary<String, Int>()

while true {
  var line = readline(fd)
  if line == "" { break }

  line = (line as NSString).stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceAndNewlineCharacterSet())
  let words = (line as NSString).componentsSeparatedByString(" ")

  for word in words as String[] {
    if (word != "") {
      if let count = wordsDict[word] {
          wordsDict[word] = count + 1
      } else {
        wordsDict[word] = 1
      }
    }
  }
}

var sortedWords = (wordsDict as NSDictionary).keysSortedByValueUsingSelector("compare:")
sortedWords = (sortedWords as NSArray).reverseObjectEnumerator().allObjects

for word in sortedWords as String[] {
  println("\(wordsDict[word]) \(word)")
}
